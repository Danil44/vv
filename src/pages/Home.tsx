import React, { Fragment, useState, useEffect, KeyboardEvent } from 'react';
import Layout from '../components/Layout';
import Header from '../components/Header';
import HomeInnerLayout from '../components/HomeInnerLayout';
import GenresMenu from '../components/GenresMenu';
import Films from '../components/Films';
import Filmer from '../models/Filmer';
import FilmProps from '../types/Film';

function Home() {
  const [genres, setGenres] = useState<string[]>([]);
  const [selectedGenre, setSelectedGenre] = useState('');
  const [films, setFilms] = useState<FilmProps[]>([]);
  const [filmsByGenre, setFilmsByGenre] = useState<FilmProps[]>([]);
  const [currentControlledView, setCurrentControlledView] = useState<
    'menu' | 'films'
  >('menu');

  useEffect(() => {
    Filmer.films.then(films => {
      setFilms(films.data.results);

      const genres = Filmer.getGenres(films.data.results);
      setGenres(genres);
      setSelectedGenre(genres[0]);
    });
  }, []);

  useEffect(() => {
    window.addEventListener('keydown', handleKeyDown as any);

    return () => window.removeEventListener('keydown', handleKeyDown as any);
  }, [currentControlledView]);

  useEffect(() => {
    setFilmsByGenre(Filmer.getFilmsByGenre(selectedGenre, films));
  }, [selectedGenre]);

  const handleKeyDown = (e: KeyboardEvent) => {
    if (e.keyCode === 66) {
      if (currentControlledView === 'menu') return;

      setCurrentControlledView('menu');
    }
  };

  const handleGenreChange = (selectedGenre: string) => {
    setSelectedGenre(selectedGenre);
  };

  const handleGenreSubmit: React.FormEventHandler<HTMLFormElement> = e => {
    e.preventDefault();
    setCurrentControlledView('films');
  };

  return (
    <Layout>
      <Fragment>
        <Header />
        <HomeInnerLayout>
          {genres.length && (
            <GenresMenu
              onGenreSubmit={handleGenreSubmit}
              onGenreChange={handleGenreChange}
              genres={genres}
              controlled={currentControlledView === 'menu'}
            />
          )}
          <Films
            controlled={currentControlledView === 'films'}
            onFilmSubmit={() => {}}
            films={filmsByGenre}
          />
        </HomeInnerLayout>
      </Fragment>
    </Layout>
  );
}

export default Home;
