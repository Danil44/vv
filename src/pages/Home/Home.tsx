import React, { Fragment, useState, useEffect, KeyboardEvent } from 'react';
import Layout from '../../components/Layout';
import Header from '../../components/Header/Header';
import HomeInnerLayout from '../../components/HomeInnerLayout';
import GenresMenu from '../../components/GenresMenu/GenresMenu';
import Films from '../../components/Films';
import Filmer from '../../models/Filmer';
import FilmProps from '../../types/Film';
import FilmView from '../../components/FilmView';
import { useRouteMatch, useHistory } from 'react-router';
import { KEY_B } from '../../hooks/useKeyboardWalking/useKeyboardWalking';

function Home() {
  const [genres, setGenres] = useState<string[]>([]);
  const [selectedGenre, setSelectedGenre] = useState('');
  const [selectedFilm, setSelectedFilm] = useState<FilmProps | null>(null);
  const [films, setFilms] = useState<FilmProps[]>([]);
  const [filmsByGenre, setFilmsByGenre] = useState<FilmProps[]>([]);
  const [currentControlledView, setCurrentControlledView] = useState<
    'menu' | 'films'
  >('menu');
  const { params } = useRouteMatch<{ filmId: string }>();
  const history = useHistory();

  const setFilm = (filmId: number) => {
    const selectedFilm = Filmer.getFilmById(films, filmId);

    if (!selectedFilm) return;

    setSelectedFilm(selectedFilm);
  };

  useEffect(() => {
    Filmer.films.then(films => {
      setFilms(films.data.results);

      const genres = Filmer.getGenres(films.data.results);
      setGenres(genres);
      setSelectedGenre(genres[0]);
    });
  }, []);

  useEffect(() => {
    if (params.filmId) {
      setFilm(+params.filmId);
    }

    const handleKeyDown = (e: KeyboardEvent) => {
      if (e.keyCode === KEY_B) {
        if (params.filmId) {
          history.push('/');
          setSelectedFilm(null);
          return setCurrentControlledView('films');
        }

        if (currentControlledView === 'menu') return;

        setCurrentControlledView('menu');
      }
    };

    window.addEventListener('keydown', handleKeyDown as any);

    return () => window.removeEventListener('keydown', handleKeyDown as any);
  }, [currentControlledView, params.filmId, setFilm]);

  useEffect(() => {
    setFilmsByGenre(Filmer.getFilmsByGenre(selectedGenre, films));
  }, [selectedGenre, films]);

  const handleGenreChange = (selectedGenre: string) => {
    setSelectedGenre(selectedGenre);
  };

  const handleGenreSubmit: React.FormEventHandler<HTMLFormElement> = e => {
    e.preventDefault();
    setCurrentControlledView('films');
  };

  const handleFilmSubmit = (filmId: number) => {
    history.push(`/${filmId}`);
    setFilm(filmId);
  };

  return (
    <Layout>
      <Fragment>
        <Header shift={Boolean(selectedFilm)} />
        <HomeInnerLayout shift={Boolean(selectedFilm)}>
          {genres.length && (
            <GenresMenu
              onGenreSubmit={handleGenreSubmit}
              onGenreChange={handleGenreChange}
              genres={genres}
              controlled={currentControlledView === 'menu' && !params.filmId}
            />
          )}
          <Films
            controlled={currentControlledView === 'films' && !params.filmId}
            onFilmSubmit={handleFilmSubmit}
            films={filmsByGenre}
          />
        </HomeInnerLayout>
        {selectedFilm ? <FilmView {...selectedFilm} /> : ''}
      </Fragment>
    </Layout>
  );
}

export default Home;
