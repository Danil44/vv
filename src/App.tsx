import React from 'react';
import Home from './pages/Home/Home';
import { Switch, Route } from 'react-router';
import { HashRouter } from 'react-router-dom';

function App() {
  return (
    <HashRouter basename="/">
      <Switch>
        <Route path="/:filmId?">
          <Home />
        </Route>
      </Switch>
    </HashRouter>
  );
}

export default App;
