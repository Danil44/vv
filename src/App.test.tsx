import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import App from './App';
import Home from './pages/Home/Home';

describe('<App />', () => {
  let wrapper: ShallowWrapper;
  beforeEach(() => (wrapper = shallow(<App />)));

  it('it has not changed', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('it should render the Home Component', () => {
    expect(wrapper.containsMatchingElement(<Home />)).toEqual(true);
  });
});
