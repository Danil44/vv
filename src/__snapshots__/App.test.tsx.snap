// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> it has not changed 1`] = `"<div class=\\"css-b07qpz-Layout\\"><header data-testid=\\"header\\" class=\\"css-1bxbtw7-Header\\"><h2 class=\\"css-1t6eos4-Header\\">FILMER</h2></header><div class=\\"css-utpa80-HomeInnerLayout\\">0<div class=\\"css-cxv9xu-Films\\"><form><input autofocus=\\"\\" class=\\"css-vxl42g\\"/></form><div class=\\"css-1hilo1s-FilmsList\\"><ul class=\\"css-1mknw5f-FilmsList\\"></ul></div></div></div></div>"`;
